### ISSUE TITLE
MSG_PEEK capability

### ISSUE THREAD

**Issue by CorvinM** [ARMmbed/mbedtls#551]
As briefly mentioned in [this discussion](https://tls.mbed.org/discussions/generic/peeking-at-data-from-an-ssl-stream-with-ssl_read) it would be helpful to have the ability to peek into available data with both ssl_read and net_recv. Adding the capability to net_recv is fairly trivial. In addition to any feedback you may have, I have a couple questions that I would like your input on:

What should the new API functions be called? these are my current thoughts:

- ssl_read4 & net_recv4 (unix style)
- ssl_read_peek & net_recv_peek
- ssl_peek & net_peek

In my opinion, the unix style functions are the best out of these because they don't limit the scope of the new function to only MSG_PEEK capability (4th argument would be a flag argument that contains the mbedtls equivalent of MSG_PEEK as well as possible others in the future)

Possible deal breaker (or possible addition as an optional feature instead): I am not positive that all the current target platforms support peeking by their network stacks.

Current status of each platform (that I know of, will update as I find out):
Linux: Yes
Windows: Yes
OS X: Yes
OpenWrt: Likely
Android: Likely
iOS: Yes
FreeRTOS: Unknown

**Comment by pjbakker** [comment]
Extending the Net module is a lot lower impact than extending the ssl interface, as it requires ABI changes (new function pointers to net functions to store)

ssl_read and ssl_write try to map to the unix functions as much as possible, so the unix style has preference on first impression to me as well. Although the purpose is less clear, so the ssl_peek() functions might useful additions as wrappers as well.

**Comment by pjbakker** [comment]
By guarding the addition with a new define (that is disabled by default), it would have no ABI impact and could be incorporated sooner.

**Comment by CorvinM** [comment]
Thanks, I was not planning on extending the peek functionality all the way down to the the transport layer for ssl_read4. Hopefully the PR makes it clear what I was trying to achieve.

**Comment by pjbakker** [comment]
It's a bit more complex than that in ssl_read4().. As decryption has already taken part on some data after a peek read...

But the gist is clear.

**Comment by mazimkhan** [comment]
@CorvinM thanks for raising the issue and making PR #563. I would like to understand the motivation behind this change. ```MSG_PEEK``` is generally used to receive a header carrying length indicator from the TCP stream and do required buffer allocations. Is this the use case behind you intend to solve?

**Comment by gilles-peskine-arm** [comment]
> Adding the capability to net_recv is fairly trivial

That's true on Unix and Windows, but I don't think it can be done on Mbed OS ([TCP](https://os.mbed.com/docs/v5.8/reference/tcpsocket.html), [UDP](https://os.mbed.com/docs/v5.8/reference/udpsocket.html)).

Instead `ssl_peek` could be implemented by doing an actual non-blocking read on the network socket and decoding the data that is available.

But I have to say that I'm not convinced that doing this in the TLS library makes sense. Why not do this on the layer above? You mention the [forum thread](https://tls.mbed.org/discussions/generic/peeking-at-data-from-an-ssl-stream-with-ssl_read)

> While it's possible to workaround this limitation by implementing a layer of code that caches bytes when you "peek" at them so they will be returned by the next call to ssl_read, this kind of workaround is clumsy at best.

Why is it  clumsier to do it in the application than in the underlying stack? This may be true if the underlying stack is storing bytes in a buffer until the application reads them: then it's relatively straightforward to let the application non-destructively read this buffer. But this is not the case with a TLS stack. Processing a TLS stream has serious side effects: an error could occur, or a TLS renegotiation. Furthermore the TLS stack does not accumulate data beyond one record and thus operates with a fixed buffer size. Reading extra network data could retrieve multiple records which needs to be stored. In contrast, if the application is peeking at upcoming data for parsing purposes, it knows how much data it wants. So it seems to me that this is best handled at the application level, without adding complexity and extra memory at the TLS layer.

**Comment by daverodgman** [comment]
This seems to be an issue for DTLS: @Thibautartis comments in #5843:

> I have implemented a DTLS server using mbedtls, I recently added the connection ID feature but I have had to resort to some ugly tricks in order to be able to process the connection ID, I'll explain
> 
> The server I implemented has a main thread listening for new incoming connections, each time it receives a packet coming from a new UDP connection it creates a thread to which it gives the ownership of the associated fd. The thread then handles the handshake and all subsequent communication with the client
> 
> In order to test the connection ID I created a basic client that connects to my server and after a while closes and reopens its mbedtls_net_context, making its DTLS packets travel from a new UDP port to the server
> 
> And here is the problem, at this point we have a packet that belongs to an existing DTLS context but coming from a new UDP connection, and I have found no way in the library to discriminate whether this packet is a new ClientHello or a connection ID'd Application Data before attempting a handshake.
> The solution I came up with for now is to try a handshake on that new UDP connection and, if it fails (which is the case here), I check for ssl.private_in_hdr[0] == MBEDTLS_SSL_MSG_CID, if this condition is satisfied then I know I can replace the old file descriptor corresponding to this connection ID by the new one and the DTLS exchanges can continue normally
> 
> The problem with this solution, other than accessing private variables of the library is that it makes the packet triggering this FD change "lost" in the sense that it has already been consumed by the library when attempting to do a handshake and cannot be retrieved in order to be processed again in the right ssl context.
> This can be a huge problem in case the packet frequency is lower than the frequency in NAT changes
> 
> This is why I think it would be necessary to be able to peek at an incoming packet, or at least be able to look for a connection ID in a packet before having it processed

**Comment by HomeACcessoryKid** [comment]
While I don't say it is a good argument, but wolfSSL_peek does exist.
Keeping parity on functions has an added value when porting between libraries...
My 2 cents
